src/charm.py

```diff
@@ -0,0 +1,250 @@
+
+logger = logging.getLogger(__name__)
+
+VAULT_STORAGE_PATH = "/srv"
+
+
+class VaultCharm(CharmBase):
+    """Main class for to handle Juju events."""
+
+    VAULT_PORT = 8200
+    VAULT_CLUSTER_PORT = 8201
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._stored.set_default(
+            root_token="", unseal_key="", root_ca="", root_token_provided=False
+        )
+        self.tls_certificates = TLSCertificatesProvides(self, "certificates")
+        self.vault = Vault(
+            url=f"http://localhost:{self.VAULT_PORT}", root_token=self._stored.root_token
+        )
+        self._service_name = self._container_name = "vault"
+        self._container = self.unit.get_container(self._container_name)
+        self.framework.observe(
+            self.tls_certificates.on.certificate_request, self._on_certificate_request
+        )
+        self.framework.observe(self.on.vault_pebble_ready, self._on_config_changed)
+        self.framework.observe(self.on.new_policy_action, self._new_policy_action)
+        self.framework.observe(self.on.new_app_role_action, self._new_app_role_action)
+        self.framework.observe(self.on.get_token_action, self._get_token_action)
+        self.framework.observe(self.on.get_root_token_action, self._get_root_token_action)
+
+    def _on_certificate_request(self, event) -> None:
+        """Handler triggered whenever there is a request made from a requirer charm to vault.
+
+        Args:
+            event: Juju event
+
+        Returns:
+            None
+        """
+        certificate = self.vault.issue_certificate(
+            cert_type=event.cert_type, common_name=event.common_name, sans=event.sans
+        )
+        self.tls_certificates.set_relation_certificate(
+            certificate=Cert(
+                cert=certificate["certificate"],
+                key=certificate["private_key"],
+                ca=certificate["issuing_ca"],
+                common_name=event.common_name,
+            ),
+            relation_id=event.relation_id,
+        )
+
+    def _on_config_changed(self, event) -> None:
+        """Handler triggerred whenever there is a config-changed event.
+
+        Args:
+            event: Juju event
+
+        Returns:
+            None
+        """
+        if not self._container.can_connect():
+            event.defer()
+            return
+        plan = self._container.get_plan()
+        layer = self._vault_layer()
+        if plan.services != layer.services:
+            self._container.add_layer(self._container_name, layer, combine=True)
+            logger.info("Added updated layer 'vault' to Pebble plan")
+            self._container.restart(self._service_name)
+            logger.info("Restarted vault service")
+
+        self._patch_storage_ownership()
+
+        if self.unit.is_leader():
+            if not self.vault.is_available:
+                logger.info("Vault is not yet available, deferring")
```

Comment:
How about setting unit status instead of logging? 
I guess in most cases, user will observe the installation process using `juju status` rather than the debug-log.

Reply:
Agreed, fixed